presto-native-execution/presto_cpp/main/operators/PartitionAndSerialize.cpp

```diff
@@ -68,19 +72,171 @@ class PartitionAndSerializeOperator : public Operator {
+      partitionVectorSize += nullsToFill;
+    }
+    return partitionVectorSize;
+  }
+
+  std::map<int32_t, int32_t> countNullsPerPartition(
+      FlatVector<int32_t>& partitionsVector) {
+    std::map<int32_t, int32_t> perPartitionNullCount;
+    const auto numInput = input_->size();
+    for (int idx = 0; idx < numInput; ++idx) {
+      if (nullRows_.isValid(idx)) {
+        perPartitionNullCount[partitionsVector.valueAt(idx)]++;
+      }
+    }
+    return perPartitionNullCount;
+  }
+
+  void fillDataWithNull(
+      FlatVector<StringView>& dataVector,
+      vector_size_t start,
+      vector_size_t end) {
+    velox::row::UnsafeRowFast emptyUnsafeRow(nullRowVector_);
+
+    auto rowSize = emptyUnsafeRow.rowSize(0);
+    auto buffer = dataVector.getBufferWithSpace(rowSize);
+    auto rawBuffer = buffer->asMutable<char>() + buffer->size();
+    buffer->setSize(buffer->size() + rowSize);
+    memset(rawBuffer, 0, rowSize);
+
+    auto emptyRowSize = emptyUnsafeRow.serialize(0, rawBuffer);
+    //  Serialize rows.
+    for (auto i = start; i < end; ++i) {
+      dataVector.setNoCopy(i, StringView(rawBuffer, emptyRowSize));
+    }
+  }
+
+  vector_size_t replicateAnyInEmptyPartitions(
+      FlatVector<int32_t>& partitionsVector,
+      const vector_size_t start) {
+    auto rawPartitions = partitionsVector.mutableRawValues();
+    const auto numInput = input_->size();
+    std::set<int32_t> nonEmptyPartitions(
+        rawPartitions, rawPartitions + numInput);
+
+    int32_t finalOutputSize = start;
+    for (auto idx = 0; idx < numPartitions_; ++idx) {
+      if (nonEmptyPartitions.find(idx) == nonEmptyPartitions.end()) {
+        rawPartitions[finalOutputSize] = idx;
+        // TODO: Update dataVector
+        finalOutputSize++;
+      }
+    }
+    return finalOutputSize;
+  }
+
+  // A partition key is considered null only if its recursive children all are
+  // nulls. collectNullRows will flatten the key columns and store the null bits
+  // in a SelectivityVector nullRows_
+  void collectNullRows() {
+    auto size = input_->size();
+    rows_.resize(size);
+    rows_.setAll();
+
+    nullRows_.resize(size);
+    nullRows_.clearAll();
+
+    decodedVectors_.resize(keyChannels_.size());
+
+    for (auto partitionKey : keyChannels_) {
+      auto& keyVector = input_->childAt(partitionKey);
+      if (keyVector->mayHaveNulls()) {
+        decodedVectors_[partitionKey].decode(*keyVector, rows_);
+        if (auto* rawNulls = decodedVectors_[partitionKey].nulls()) {
+          bits::orWithNegatedBits(
+              nullRows_.asMutableRange().bits(), rawNulls, 0, size);
+        }
+      }
+    }
+    nullRows_.updateBounds();
+  }
+
   RowVectorPtr getOutput() override {
     if (!input_) {
       return nullptr;
     }
+    RowVectorPtr reorderedInputVector = reorderInputsIfNeeded();
+    velox::row::UnsafeRowFast unsafeRow(reorderedInputVector);
+    if (!nullRowVector_) {
+      nullRowVector_ =
```

Comment:
This is incorrect. We shouldn't pass a null const vector but pass a vector with null child? And verified locally.